mlir/include/mlir/Dialect/Ptr/IR/PtrOps.td

```diff
@@ -17,6 +17,75 @@ include "mlir/Interfaces/SideEffectInterfaces.td"
 include "mlir/Interfaces/ViewLikeInterface.td"
 include "mlir/IR/OpAsmInterface.td"
 
+//===----------------------------------------------------------------------===//
+// FromPtrOp
+//===----------------------------------------------------------------------===//
+
+def Ptr_FromPtrOp : Pointer_Op<"from_ptr", [
+    Pure, OptionalTypesMatchWith<"metadata type", "result", "metadata",
+            "PtrMetadataType::get(cast<PtrLikeTypeInterface>($_self))">
+  ]> {
+  let summary = "Casts a `!ptr.ptr` value to a ptr-like value.";
+  let description = [{
+    The `from_ptr` operation casts a `ptr` value to a ptr-like object. It's
+    important to note that:
+    - The ptr-like object cannot be a `!ptr.ptr`.
+    - The memory-space of both the `ptr` and ptr-like object must match.
+    - The cast is side-effect free.
+
+    If the ptr-like object type has metadata, then the operation expects the
+    metadata as an argument or expects that the flag `trivial_metadata` is set.
+    If `trivial_metadata` is set, then it is assumed that the metadata can be
+    reconstructed statically from the pointer-like type.
+
+    Example:
+
+    ```mlir
+    %typed_ptr = ptr.from_ptr %ptr : !ptr.ptr<0> -> !my.ptr<f32, 0>
+    %memref = ptr.from_ptr %ptr metadata %md : !ptr.ptr<0> -> memref<f32, 0>
+    %memref = ptr.from_ptr %ptr trivial_metadata : !ptr.ptr<0> -> memref<f32, 0>
+    ```
+  }];
+
+  let arguments = (ins Ptr_PtrType:$ptr,
+                       Optional<Ptr_PtrMetadata>:$metadata,
+                       UnitProp:$hasTrivialMetadata);
+  let results = (outs PtrLikeTypeInterface:$result);
+  let assemblyFormat = [{
+    $ptr (`metadata` $metadata^)? (`trivial_metadata` $hasTrivialMetadata^)?
+    attr-dict `:` type($ptr) `->` type($result)
+  }];
+  let hasFolder = 1;
+  let hasVerifier = 1;
+}
+
+//===----------------------------------------------------------------------===//
+// GetMetadataOp
+//===----------------------------------------------------------------------===//
+
+def Ptr_GetMetadataOp : Pointer_Op<"get_metadata", [
```

Comment:
Is this operation similar to `memref.extract_strided_metadata`? Are we still going to need `memref.extract_strided_metadata` or this a replacement?

Reply:
It's somewhat similar. They are both extracting the information. The difference is that `ptr` doesn't make assumptions of what information might be.

I would argue for removal of `extract_strided_metadata`. But that requires further discussion.

Reply:
I'm not talking about replacement, but evaluate the removal. Because, that op is clearly tied to the lowering structure to `LLVM` or from upstream.

I could reimplement things and say `%base` is not part of the descriptor anymore, because the `memref` type doesn't specify doesn't talk about its implementation.

Now, if we remove `extract_strided_metadata`, I would argue we add `get_sizes`, `get_strides`. But all of that is a separate discussion.

Reply:
> This one is opaque, while memref.extract_strided_metadata isn't. I don't quite see how this one can replace memref.extract_strided_metadata in general?

This makes sense. I'm just trying to see what you can do with `get_metadata` and what the interactions between the `memref` dialect and the `ptr` dialect could be.